### ISSUE TITLE
PostgreSQL UDT not working when used as stored procedure return type

### ISSUE THREAD

**Issue by cardamon** [jOOQ/jOOQ#6819]
### Expected behavior and actual behavior:
Expected: UDT working properly when used as return type in a PostgreSQL stored procedure
Actual:
```
org.jooq.exception.DataAccessException: SQL [select "jooq_bug_test" from "jooq_bug_test"("some_id" := ?)]; Error while reading field: "jooq_bug_test", at JDBC index: 1
	at org.jooq_3.10.1.POSTGRES_9_5.debug(Unknown Source)
	at org.jooq_3.10.1.POSTGRES_9_5.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:2199)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:701)
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1619)
	at org.jooq.impl.CursorImpl$CursorIterator.hasNext(CursorImpl.java:1570)
	at org.jooq.impl.CursorImpl.fetchNext(CursorImpl.java:392)
	at org.jooq.impl.CursorImpl.fetch(CursorImpl.java:378)
	at org.jooq.impl.CursorImpl.fetch(CursorImpl.java:289)
	at org.jooq.impl.AbstractResultQuery.execute(AbstractResultQuery.java:288)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:344)
	at org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:314)
	at org.jooq.impl.SelectImpl.fetch(SelectImpl.java:2594)
	at org.jooq.impl.AbstractRoutine.executeSelectFromPOSTGRES(AbstractRoutine.java:430)
	at org.jooq.impl.AbstractRoutine.execute(AbstractRoutine.java:332)
	at org.jooq.impl.AbstractRoutine.execute(AbstractRoutine.java:314)
	(<cut>) Routines.jooqBugTest(Routines.java:251)
	(<cut>)
Caused by: java.sql.SQLException: Error while reading field: "jooq_bug_test", at JDBC index: 1
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1714)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.operate(CursorImpl.java:1674)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.operate(CursorImpl.java:1639)
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:122)
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1603)
	... 69 more
Caused by: org.jooq.exception.DataTypeException: Error while creating array
	at org.jooq.impl.DefaultBinding.pgNewArray(DefaultBinding.java:2545)
	at org.jooq.impl.DefaultBinding.pgFromString(DefaultBinding.java:2376)
	at org.jooq.impl.DefaultBinding.pgSetValue(DefaultBinding.java:2550)
	at org.jooq.impl.DefaultBinding$2.operate(DefaultBinding.java:2440)
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:122)
	at org.jooq.impl.DefaultBinding.pgNewRecord(DefaultBinding.java:2433)
	at org.jooq.impl.DefaultBinding.get(DefaultBinding.java:1633)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1704)
	... 73 more
Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:657)
	at java.util.ArrayList.get(ArrayList.java:433)
	at org.jooq.impl.DefaultBinding$2.operate(DefaultBinding.java:2440)
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:122)
	at org.jooq.impl.DefaultBinding.pgNewRecord(DefaultBinding.java:2433)
	at org.jooq.impl.DefaultBinding.pgFromString(DefaultBinding.java:2387)
	at org.jooq.impl.DefaultBinding.pgFromString(DefaultBinding.java:2285)
	at org.jooq.impl.DefaultBinding.pgNewArray(DefaultBinding.java:2539)
	... 80 more
```
Could this be related to https://github.com/jOOQ/jOOQ/issues/5613 ?
### Steps to reproduce the problem:
- Create a stored procedure as follows. Probably not a minimal example, but one that closely resembles our use case:
```sql
DROP TYPE IF EXISTS
    my_info,
    my_interval
CASCADE;

CREATE TYPE my_interval AS (
    period_start timestamptz,
    period_end timestamptz,
    errors text[]
);

CREATE TYPE my_info AS (
    some_id bigint,
    invals my_interval[]
);

CREATE OR REPLACE FUNCTION jooq_bug_test(
    some_id bigint
) RETURNS my_info AS
$body$
DECLARE
    result my_info;
    ival1 my_interval;
    ival2 my_interval;
BEGIN
    ival1 := row('2017-11-10T00:00:00+01:00', '2017-11-11T00:00:00+01:00', ARRAY['foo', 'bar', 'baz']);
    ival2 := row('2017-11-11T00:00:00+01:00', '2017-11-12T00:00:00+01:00', ARRAY['lorem', 'ipsum']);

    result.some_id := some_id;
    result.invals := ARRAY[ival1, ival2];

    RAISE NOTICE 'result: %', result;
    RETURN result;
END
$body$
LANGUAGE plpgsql;
```
* Next, with the code generated by jOOQ, do something like the following:
```java
System.out.format(
                "The record mapped into a pojo: %s%n",
                Routines
                        .jooqBugTest(DSL.using(dataSource, SQLDialect.POSTGRES_9_5, settings).configuration(), 12356L)
                        .into(MyInfoPojo.class));
```

### Versions:

- jOOQ: 3.10.1
- Java:
```
$ java -version
openjdk version "1.8.0_151"
OpenJDK Runtime Environment (build 1.8.0_151-8u151-b12-0ubuntu0.16.04.2-b12)
OpenJDK 64-Bit Server VM (build 25.151-b12, mixed mode)
```
- Database (include vendor):
```
$ dpkg-query -l | grep postgresql-9.6
ii  postgresql-9.6                             9.6.3-0ubuntu0.17.04
```
- JDBC Driver (include name if inofficial driver):
```xml
<dependency>
    <groupId>org.postgresql</groupId>
    <artifactId>postgresql</artifactId>
    <version>42.1.4</version>
</dependency>
```

**Comment by cardamon** [comment]
Oh BTW when executing
```sql
SELECT jooq_bug_test(123);
```
this is the result:
```
(123,"{""(\\""2017-11-10 00:00:00+01\\"",\\""2017-11-11 00:00:00+01\\"",\\""{\\""\\""foo, bar, baz\\""\\""}\\"")"",""(\\""2017-11-11 00:00:00+01\\"",\\""2017-11-12 00:00:00+01\\"",\\""{\\""\\""lorem, ipsum\\""\\""}\\"")""}")
```
So I guess that's not being parsed correctly by jOOQ?

**Comment by lukaseder** [comment]
Thank you very much for the detailed report - and sorry for the delay. I'll look into this right away

**Comment by lukaseder** [comment]
Hmm, it seems as though `PostgresUtils.toPGObjectOrArray()` makes a mistake parsing this serialised object form:

```
{"(\"2017-11-10 00:00:00+01\",\"2017-11-11 00:00:00+01\",\"{foo,bar,baz}\")","(\"2017-11-11 00:00:00+01\",\"2017-11-12 00:00:00+01\",\"{lorem,ipsum}\")"}
```

**Comment by lukaseder** [comment]
The problem is here:
https://github.com/jOOQ/jOOQ/blob/version-3.10.1/jOOQ/src/main/java/org/jooq/util/postgres/PostgresUtils.java#L369

That first backslash is not parsed correctly

**Comment by lukaseder** [comment]
The error is here:
https://github.com/jOOQ/jOOQ/blob/version-3.10.1/jOOQ/src/main/java/org/jooq/util/postgres/PostgresUtils.java#L369

That first backslash is not escaping the double quote

**Comment by lukaseder** [comment]
This will need some more thought. It looks as though the escaping patterns aren't as straightforward (again) as I hoped. So, this won't make it into tomorrows 3.10.2 anymore, I'm afraid.

**Comment by cardamon** [comment]
Thanks for picking this up :+1:

**Comment by lukaseder** [comment]
Same issue: https://github.com/jOOQ/jOOQ/issues/10467

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.14.0. Backport scheduled for 3.13.5 (#10471)
